src/iceberg/table_scan.h

```diff
@@ -0,0 +1,200 @@
+
+  /// \brief Sets whether the scan should be case-sensitive.
+  /// \param case_sensitive Whether the scan is case-sensitive.
+  /// /return Reference to the builder.
+  TableScanBuilder& WithCaseSensitive(bool case_sensitive);
+
+  /// \brief Sets an option for the scan.
+  /// \param property The name of the option.
+  /// \param value The value of the option.
+  /// \return Reference to the builder.
+  TableScanBuilder& WithOption(std::string property, std::string value);
+
+  /// \brief Sets an optional limit on the number of rows to scan.
+  /// \param limit Optional limit on the number of rows.
+  /// \return Reference to the builder.
+  TableScanBuilder& WithLimit(std::optional<int64_t> limit);
+
+  /// \brief Builds and returns a TableScan instance.
+  /// \return A Result containing the TableScan or an error.
+  Result<std::unique_ptr<TableScan>> Build();
+
+ private:
+  /// \brief the file I/O instance for reading manifests and data files.
+  std::shared_ptr<FileIO> file_io_;
+  /// \brief column names to project in the scan.
+  std::vector<std::string> column_names_;
+  /// \brief snapshot ID to scan, if specified.
+  std::optional<int64_t> snapshot_id_;
+  /// \brief Context for the scan, including snapshot, schema, and filter.
+  TableScanContext context_;
+};
+
+/// \brief Represents a configured scan operation on a table.
+class ICEBERG_EXPORT TableScan {
+ public:
+  virtual ~TableScan() = default;
+
+  /// \brief Constructs a TableScan with the given context and file I/O.
+  /// \param context Scan context including snapshot, schema, and filter.
+  /// \param file_io File I/O instance for reading manifests and data files.
+  TableScan(TableScanContext context, std::shared_ptr<FileIO> file_io);
+
+  /// \brief Returns the snapshot being scanned.
+  /// \return A shared pointer to the snapshot.
+  const std::shared_ptr<Snapshot>& snapshot() const;
+
+  /// \brief Returns the projected schema for the scan.
+  /// \return A shared pointer to the projected schema.
+  const std::shared_ptr<Schema>& projection() const;
+
+  /// \brief Returns the scan context.
+  /// \return A reference to the TableScanContext.
+  const TableScanContext& context() const;
+
+  /// \brief Returns the file I/O instance used for reading manifests and data files.
+  /// \return A shared pointer to the FileIO instance.
+  const std::shared_ptr<FileIO>& io() const;
+
+  /// \brief Plans the scan tasks by resolving manifests and data files.
+  /// \return A Result containing scan tasks or an error.
+  virtual Result<std::vector<std::shared_ptr<FileScanTask>>> PlanFiles() const = 0;
+
+ protected:
+  /// \brief context for the scan, including snapshot, schema, and filter.
+  const TableScanContext context_;
+  /// \brief File I/O instance for reading manifests and data files.
+  std::shared_ptr<FileIO> file_io_;
+};
+
+/// \brief A scan that reads data files and applies delete files to filter rows.
+class ICEBERG_EXPORT DataScan : public TableScan {
```

Comment:
```suggestion
class ICEBERG_EXPORT DataTableScan : public TableScan {
```

I think `DataScan` is ambiguous and people may assume that it will return rows of data instead of the file list. `DataTable` is better to differentiate from `MetadataTable`.